spec.html

```diff
@@ -20665,6 +21279,230 @@ <h1>Static Semantics: PrototypePropertyNameList</h1>
       </emu-alg>
     </emu-clause>
 
+    <emu-clause id="sec-class-definitions-static-semantics-allprivateidentifiersvalid" type="sdo" aoid="AllPrivateIdentifiersValid">
+      <h1>Static Semantics: AllPrivateIdentifiersValid</h1>
+      <p>With parameter _names_.</p>
+      <emu-grammar>MemberExpression : MemberExpression `.` PrivateIdentifier</emu-grammar>
+      <emu-alg>
+        1. If StringValue of |PrivateIdentifier| is in _names_, return *true*.
+        1. Return *false*.
+      </emu-alg>
+
+      <emu-grammar>CallExpression : CallExpression `.` PrivateIdentifier</emu-grammar>
+      <emu-alg>
+        1. If StringValue of |PrivateIdentifier| is in _names_, return *true*.
+        1. Return *false*.
+      </emu-alg>
+
+      <emu-grammar>OptionalChain[Yield, Await] : `?.` PrivateIdentifier</emu-grammar>
+      <emu-alg>
+        1. If StringValue of |PrivateIdentifier| is in _names_, return *true*.
+        1. Return *false*.
+      </emu-alg>
+
+      <emu-grammar>OptionalChain[Yield, Await] : OptionalChain[?Yield, ?Await] `.` PrivateIdentifier</emu-grammar>
+      <emu-alg>
+        1. If StringValue of |PrivateIdentifier| is in _names_, return *true*.
+        1. Return *false*.
+      </emu-alg>
+
+      <emu-grammar>ClassBody : ClassElementList</emu-grammar>
+      <emu-alg>
+        1. Let _newNames_ be a copy of _names_.
+        1. Append to _newNames_ the elements of PrivateBoundIdentifiers of |ClassBody|.
+        1. Return AllPrivateIdentifiersValid of |ClassElementList| with argument _newNames_.
+      </emu-alg>
+
+      <p>For all other grammatical productions, recurse on subexpressions/substatements, passing in the _names_ of the caller. If all pieces return *true*, then return *true*. If any returns *false*, return *false*.</p>
```

Comment:
This is not as precise as I'd like. Could we give roughly the same treatment as `Contains`? Something like


```
Every grammar production alternative in this specification which is not listed below implicitly has the following default definition of AllPrivateIdentifiersValid:

1. For each child node _child_ of this Parse Node, do
  a. If child is an instance of a nonterminal, then
    i. If AllPrivateIdentifiersValid of _child_ with argument _names_ is *false*, return *false*.
2. Return *true*.
```

Reply:
`If _child_ is`